docs/user-guide/content-type.md

```diff
@@ -0,0 +1,232 @@
+
+The `np` content type will decode / encode V2 payloads to a NumPy Array, taking
+into account the following:
+
+- The `datatype` field will be matched to the closest [NumPy
+  `dtype`](https://numpy.org/doc/stable/reference/arrays.dtypes.html).
+- The `shape` field will be used to reshape the flattened array expected by the
+  V2 protocol into the expected tensor shape.
+
+When using the NumPy Array content type at the request-level, it will decode
+the entire request by considering only the first `input` element.
+This can be used as a helper for models which only expect a single tensor.
+
+### Pandas DataFrame
+
+```{note}
+The `pd` content type can be _stacked_ with other content types.
+This allows the user to use a different set of content types to decode each of
+the columns.
+```
+
+The `pd` content type will decode / encode a V2 request into a Pandas
+DataFrame.
+For this, it will expect that the DataFrame is shaped in a **columnar way**.
+That is,
+
+- Each entry of the `inputs` list (or `outputs`, in the case of
+  responses), will represent a column of the DataFrame.
+- Each of these entires, will contain all the row elements for that particular
+  column.
+- The `shape` field of each `input` (or `output`) entry will contain (at least)
+  the amount of rows included in the dataframe.
+
+For example, if we consider the following dataframe:
+
+| A   | B   | C   |
+| --- | --- | --- |
+| a1  | b1  | c1  |
+| a2  | b2  | c2  |
+| a3  | b3  | c3  |
+| a4  | b4  | c4  |
+
+We could encode it to the V2 Inference Protocol as the following payload:
+
+```{code-block} json
+---
+emphasize-lines: 3, 7-8, 13-14, 19-20
+---
+{
+  "parameters": {
+    "content_type": "pd"
+  },
+  "inputs": [
+    {
+      "name": "A",
+      "data": ["a1", "a2", "a3", "a4"]
+      "datatype": "BYTES",
+      "shape": [3],
+    },
+    {
+      "name": "B",
+      "data": ["b1", "b2", "b3", "b4"]
+      "datatype": "BYTES",
+      "shape": [3],
+    },
+    {
+      "name": "C",
+      "data": ["c1", "c2", "c3", "c4"]
+      "datatype": "BYTES",
+      "shape": [3],
+    },
+  ]
+}
+```
+
+### UTF-8 String
+
+The `str` content type lets you encode / decode a V2 input into a UTF-8
+Python string, taking into account the following:
+
+- The expected `datatype` is `BYTES`.
+- The `shape` field represents the number of "strings" that are encoded in
+  the payload (e.g. the `["hello world", "one more time"]` payload will have a
+  shape of 2 elements).
+
+When using the `str` content type at the request-level, it will decode the
+entire request by considering only the first `input` element.
+This can be used as a helper for models which only expect a single string or of
+set of strings.
+
+### Base64
+
+The `base64` content type will decode a binary V2 payload into a Base64-encoded
+string (and viceversa), taking into account the following:
+
+- The expected `datatype` is `BYTES`.
+- The `data` field should contain the **raw bytes** (i.e. before encoding them
```

Comment:
whats the reason it doesnt come in base64? it is because json will serialise/deserialise anyway in base64?

Reply:
This is mainly thinking of protobuf, where sending the payload as bytes would be more efficient. You're right though in that, when using JSON, probably sending the actual base64 could be better.

Having said that, I'd say that changing the codecs is outside the scope of this PR, which only aims to document them.

Reply:
Just had a look, and it actually expects the base64-encoded arrays, and then it decodes them into binaries.